c2rust-analyze/src/main.rs

```diff
@@ -321,6 +321,46 @@ where
     }
 }
 
+fn mark_foreign_fixed<'tcx>(
+    gacx: &mut GlobalAnalysisCtxt<'tcx>,
+    gasn: &mut GlobalAssignment,
+    tcx: TyCtxt<'tcx>,
+) {
+    // FIX the inputs and outputs of function declarations in extern blocks
+    for did in gacx.foreign_mentioned_tys.clone() {
+        if let DefKind::Fn = tcx.def_kind(did) {
+            let sig = tcx.erase_late_bound_regions(tcx.fn_sig(did));
+            let inputs = sig
+                .inputs()
+                .iter()
+                .map(|&ty| gacx.assign_pointer_ids_with_info(ty, PointerInfo::ANNOTATED))
+                .collect::<Vec<_>>();
+            for input in inputs {
+                make_ty_fixed(gasn, input);
+            }
+
+            let output = gacx.assign_pointer_ids_with_info(sig.output(), PointerInfo::ANNOTATED);
+            make_ty_fixed(gasn, output)
```

Comment:
This loop creates a bunch of `LTy`s and marks them `FIXED` but doesn't store them anywhere, so they have no effect on anything.  `inputs` and `output` should probably be bundled into an `LFnSig` and then stored in either `gacx.fn_sigs` or some new map.

Reply:
`gacx.fn_sigs` just stores `LFnSig`s for body owners, so no body-less `extern fn`s (how would some like a non-default `trait` `fn` be handled, by the way?) are included.  Doesn't that mean we would never rewrite such an `extern fn` in the first place?

Reply:
We shouldn't rewrite `extern`-block `fn`s, but we need to retrieve their signatures anywhere they're called.  That's why I was thinking it might be reasonable to put them in `fn_sigs`, alongside the signatures of all the normal `fn`s.

Reply:
Don't we just currently crash on calls to `extern fn`s?  That's why I'm working on fixing with hardcoding permissions for `libc` `fn`s, but that's probably going to work a bit differently from `fn_sigs`.  Right now, and for any `extern fn` we don't hardcode, we'll crash on the call when we try to lookup the `LFnSig` from `fn_sigs`.  I just want to make sure we don't silently do the wrong thing instead of crashing if we include these `extern fn`s in `fn_sigs`.

Reply:
I've added them to `gacx.fn_sigs` in 7f8bddbe

Reply:
> Don't we just currently crash on calls to `extern fn`s? That's why I'm working on fixing with hardcoding permissions for `libc` `fn`s

Yes - there are two parts to libc function handling:
1. For analysis, we need to know the `READ`/`WRITE`/`OFFSET` permissions on the function signature.  This is the part involving hard-coded permissions.
2. For rewriting, we need to generate casts around calls to libc functions.  Setting `FIXED` on the signature triggers insertion of these casts.



> Right now, and for any `extern fn` we don't hardcode, we'll crash on the call when we try to lookup the `LFnSig` from `fn_sigs`. I just want to make sure we don't silently do the wrong thing instead of crashing if we include these `extern fn`s in `fn_sigs`.

Good point.  Currently `util::ty_callee` checks for `extern`-block `fn`s directly (by looking at the parent `DefKind`), so that part will still work after this change.  But are there other places in the code that assume `extern`-block `fn`s will be absent from `fn_sigs`?

Reply:
@aneksteind @spernsteiner.  I could move the foreign `fn` sigs to a new map, right?  As long as I handle it during calls (which should be quite simple as I'm already changing that stuff), it should be fine, right?  I think it can simplify things with #980 and #999.

Reply:
I'm not sure how that helps, but yeah, you could move them as long as you update all the places where calls are handled (which includes at least `dataflow::type_check` and `rewrite::expr::mir_op` - I don't think `borrowck::type_check` handles calls yet).